### PR TITLE
Public Chat Message Signing

### PR DESCRIPTION
**Pull request by Mikunj**

### CODE REVIEW

File: java/src/main/java/org/whispersystems/signalservice/loki/api/LokiAPI.kt [hunk 1]
```diff
@@ -40,6 +40,7 @@ class LokiAPI(private val userHexEncodedPublicKey: String, private val database:
     // region Types
     sealed class Error(val description: String) : Exception() {
         object Generic : Error("An error occurred.")
+        object SigningFailed: Error("Signing failed.")
```
Comment (gmbnt) [inline java/src/main/java/org/whispersystems/signalservice/loki/api/LokiAPI.kt]: The description should provide more information than the error name does

File: java/src/main/java/org/whispersystems/signalservice/loki/api/LokiGroupChatAPI.kt [hunk 2]
```diff
@@ -65,22 +65,22 @@ public class LokiGroupChatAPI(private val userHexEncodedPublicKey: String, priva
                     200 -> {
                         try {
                             val bodyAsString = response.body()!!.string()
-                            @Suppress("NAME_SHADOWING") val body = JsonUtil.fromJson(bodyAsString, Map::class.java)
-                            val base64EncodedChallenge = body["cipherText64"] as String
+                            val root = JsonUtil.fromJson(bodyAsString)
+                            val base64EncodedChallenge = root.get("cipherText64").asText()
```
Comment (nielsandriesse) [inline java/src/main/java/org/whispersystems/signalservice/loki/api/LokiGroupChatAPI.kt:69]: If you do this you also need to check that the value returned by `asText()` is not empty. Currently the way this code works is that it relies on an error to be thrown if the JSON being parsed is invalid. This change breaks that pattern.

File: java/src/main/java/org/whispersystems/signalservice/loki/api/LokiGroupMessage.kt [hunk 3]
```diff
@@ -1,37 +1,96 @@
...
         public val quoteeHexEncodedPublicKey: String,
-        public val quotedMessageBody: String
+        public val quotedMessageBody: String,
+        public val quotedMessageServerId: Long? = null
     ) {
-
-        internal fun toJSON(): String {
-            return "{ \"id\" : $quotedMessageTimestamp, \"author\" : \"$quoteeHexEncodedPublicKey\", \"text\" : \"$quotedMessageBody\" }"
+        internal fun jsonMap(): Map<String, Any> {
+            return sortedMapOf("id" to quotedMessageTimestamp, "author" to quoteeHexEncodedPublicKey, "text" to quotedMessageBody)
```
Comment (gmbnt) [inline java/src/main/java/org/whispersystems/signalservice/loki/api/LokiGroupMessage.kt]: Why does this need to be a `SortedMap`?

File: java/src/main/java/org/whispersystems/signalservice/loki/api/LokiGroupChatAPI.kt (hunk 2 above)
Comment (Mikunj) [reply java/src/main/java/org/whispersystems/signalservice/loki/api/LokiGroupChatAPI.kt:69]: Right but it should be fine because `decrypt` would fail anyway since the data is empty. Also from my testing, if the field `cipherText64` didn't exist then it would throw an error, it would only return an empty string if `cipherText64` was set to null.
